### ISSUE TITLE
image in multiple deployments built multiple times

### ISSUE THREAD

**Issue by nicks** [tilt-dev/tilt#2992]
In the k8s tilt channel, jasonmay observes:
> if I have a Tilt image used by multiple deployments, it builds for all deployments on start, running multiple unnecessary builds of the same image on a trigger.

These builds should *generally* be fast due to docker build caching, but there are times when that caching falls down, and there's still some build overhead for a completely cached build (e.g., tarring up the docker build context)

Ideally, Tilt would only build the image once.

This comes up particularly in airflow-based workflows, where the same image ends up in 4 places.

**Comment by landism** [comment]
> These builds should generally be fast due to docker build caching

Reading this, it's not clear if it's saying that we expect ~all builds in tilt to be fast due to docker build caching, or if it's saying that the first time we build the image in a run it will be normal, and subsequent builds should be fast due to caching.

I think the former is untrue, and the second misunderstands the issue:

The issue AIUI it is: if I have three deployments that all use the same image and start tilt, tilt will try to build those three resources, and each one will perform the exact same build steps.

Ensuring Tilt only builds the image once is a good solution, but maybe expensive. A couple of other options to this version of the problem that may be cheaper:
* ensure resources that share images don't build in parallel
* allow users to specify that certain resources should never build at the same time (might also be useful for more granular resolution of some of the local resource shenanigans than disabling parallel builds of local resources)

**Comment by nicks** [comment]
ya, those are good points!

**Comment by nicks** [comment]
@landism do you think we should file a separate issue for the parallelism problem, or bundle it with this one?

**Comment by maiamcc** [comment]
Is #2855 a subset of this issues? Or sufficiently different that they should remain as separate issues.

**Comment by nicks** [comment]
they're slightly different. This is particularly focused on the airflow use-case (where you have 1 image in 4 different deployments), whereas that one is particularly focused on the base-image use case (where you have 4 images depending on 1 base image)/ I could imagine a fix for one not fixing the others.

**Comment by nicks** [comment]
OK, with the latest release, the scheduler won't try to parallelize two deployments that have the same image, but will still repeat the builds. I'm going to leave this open for the general case of skipping the redundant builds entirely.

**Comment by crhayes** [comment]
I believe I am running into a similar issue using Tilt with Docker Compose.

Because of certain constraints of ours, I have one image with all dependencies and built source code. My `docker-compose.yml` file specifies multiple services that spin up containers from this one image.

I am building this one image with a  `custom_build`, however `custom_build` is called once for each service with a unique `$EXPECTED_REF`.

What I am attempting to do is have the first call to `custom_build` build the image, then on subsequent calls do a check to see if the image `$EXPECTED_REF` exists and just exit early, which would prevent trying to build the same image multiple times. However, this is not possible because  the `$EXPECTED_REF` is unique for each service defined in `docker-compose.yml`.

Although the layers are cached on subsequent builds, I'm finding it's still significantly slower than just not doing the build at all (especially with ~ a dozen services).

**Comment by jazzdan** [comment]
Here's an isolated repro case for this https://github.com/jazzdan/one-image-multiple-deployments

**Comment by kollender** [comment]
+1 for that.

**Comment by kollender** [comment]
In 0.12.4 the docker build built images in parallel, this is not the ideal as I expect it to be (building just a single image instead of image per deployment), but it is better for us - it is much faster than now.
We don't want to work with 0.12.4, but in the current version - let's say we want to work with tilt on 10 deployments and using some remote build - it takes docker build time (let's say 3 minutes) multiple 10, just to start using tilt.
The ideal is to build the docker once, and use the same image in all targeted deployments.
It would be great if you could prioritize it, especially nowadays when we are working from home (coronavirus) and trying to reduce every second we can due to internet speed limits.

**Comment by nicks** [comment]
as of https://github.com/tilt-dev/tilt/pull/3362, Tilt will no longer build the same image multiple times in this case

But it will still build them in serial. I filed a separate issue for the parallelization: https://github.com/tilt-dev/tilt/issues/3468

**Comment by nicks** [comment]
unfortunately, we had to roll the fix for this back :sob:

**Comment by nicks** [comment]
rolled forward a fix in https://github.com/tilt-dev/tilt/pull/3553 that i feel a lot better about
